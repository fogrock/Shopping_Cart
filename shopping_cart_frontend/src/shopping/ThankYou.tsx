import Typography from '@mui/material/Typography';

export function ThankYou() {
    return (
        <div>
            <Typography variant="h3" gutterBottom component="div">
                Thank you.
            </Typography>
            <Typography variant="h5" gutterBottom component="div">
                Your order was completed successfully
            </Typography>
            <Typography variant="button" display="block" gutterBottom>
                order id: {sessionStorage.getItem('lastOrderId')}
            </Typography>
        </div>
    );
}