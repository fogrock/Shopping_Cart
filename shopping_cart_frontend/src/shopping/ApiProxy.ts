import axios from "axios";
import { Cart } from "./CartApi";

const baseUrl: string = "https://localhost:7267";

export interface MoneyAmount {
    amount: number;
    currencyCode: string;
}

export interface Product {
    productId: number,
    productName: string,
    productDescription: string,
    price: MoneyAmount
}

export class CartUpdateRequest {
    cart: Cart;
    countryName: string;
    constructor(cart: Cart, countryName: string) {
        this.cart = cart;
        this.countryName = countryName;
    }
}

export class Order {
    orderId: string;
    orderIsoDate: string;
    cart: Cart;
    countryName: string;
    constructor(orderId: string, orderIsoDate: string, cart: Cart, countryName: string) {
        this.orderId = orderId;
        this.orderIsoDate = orderIsoDate;
        this.cart = cart;
        this.countryName = countryName;
    }
}

export class ApiProxy {

    getProductList(countryName: string): Promise<Product[]> {
        return axios.get(`${baseUrl}/Product`, {
            params: {
                countryName: countryName
            }
        })
            .then(result => result.data);
    }

    getCountryList(): Promise<string[]> {
        return axios.get(`${baseUrl}/Country`)
            .then(result => result.data);
    }

    updateCartModel(request: CartUpdateRequest): Promise<Cart> {
        return axios.post(`${baseUrl}/CartUpdate`, request)
            .then(result => { return result.data; });
    }

    getPostage(subTotal: number, countryName: string): Promise<MoneyAmount> {
        return axios.get(`${baseUrl}/Postage`, {
            params: {
                subTotal: subTotal,
                countryName: countryName
            }
        })
            .then(result => result.data);
    }

    placeOrder(order: Order): Promise<boolean> {
        return axios.post(`${baseUrl}/PlaceOrder`, order)
            .then(result => { return result.data; });
    }
}