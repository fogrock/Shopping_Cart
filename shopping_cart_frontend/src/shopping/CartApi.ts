import { Product } from "./ApiProxy";

export class CartItem {
    product: Product;
    count: number;

    constructor(product: Product, count: number) {
        this.product = product;
        this.count = count;
    }
}

export class Cart {
    cartItems: CartItem[];
    constructor(cartItems: CartItem[]) {
        this.cartItems = cartItems;
    }

    saveCart(): void {
        sessionStorage.setItem('shoppingCart', JSON.stringify(this.cartItems));
    }

    addItemToCart(product: Product, count: number): void {
        const item: CartItem | undefined = this.cartItems.find(item => item.product.productId === product.productId && item.product.productName === product.productName);
        if (item != undefined) {
            item.count++;
        }
        else {
            const newItem: CartItem = new CartItem(product, count);
            this.cartItems.push(newItem);
        }
        this.saveCart();
    }

    removeItemFromCartAll(product: Product) {
        this.cartItems.map
            (item => {
                if (item.product.productId === product.productId && item.product.productName === product.productName) {
                    this.cartItems.splice(this.cartItems.indexOf(item), 1);
                    this.saveCart();
                    return;
                };
            });
    }
}

export class CartApi {
    loadCart(): Cart {
        const cart: Cart = new Cart([]);
        const cartString: string | null = sessionStorage.getItem('shoppingCart')
        if (typeof cartString === "string") cart.cartItems = JSON.parse(cartString);
        cart.cartItems.sort(function (a, b) {
            const productName1: string = a.product.productName.toUpperCase();
            const productName2: string = b.product.productName.toUpperCase();
            return (productName1 < productName2) ? -1 : (productName1 > productName2) ? 1 : 0;
        });
        return cart;
    }

    saveNewCart(cart: Cart): void {
        sessionStorage.setItem('shoppingCart', JSON.stringify(cart.cartItems));
    }
}