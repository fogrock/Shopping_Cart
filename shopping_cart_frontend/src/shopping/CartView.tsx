import { useEffect } from "react";
import { ApiProxy, CartUpdateRequest, MoneyAmount, Order } from "./ApiProxy";
import { Cart, CartApi, CartItem } from "./CartApi";
import { AgGridReact } from 'ag-grid-react';
import { AgGridEvent, GridApi } from 'ag-grid-community';
import 'ag-grid-community/dist/styles/ag-grid.css';
import 'ag-grid-community/dist/styles/ag-theme-alpine.css';
import { createGuid, formatNumber } from '../shared/Utils';
import Alert from '@mui/material/Alert';
import React from "react";
import Button from '@mui/material/Button';
import '../App.css';

export interface CartProps {
    countryName: string;
    onOrderProcessed: () => void;
}

export function CartView(props: CartProps) {
    const apiProxy = new ApiProxy();
    const cartApi = new CartApi();
    const loadedCart = cartApi.loadCart();
    const { countryName, onOrderProcessed } = props;
    const [error, setError] = React.useState<string | undefined>(undefined);
    const [postage, setPostage] = React.useState<MoneyAmount | undefined>(undefined);
    const [selectedProducts, setSelectedProducts] = React.useState<number[]>([]);
    const [gridApi, setGridApi] = React.useState<GridApi | undefined>(undefined);
    const cartRef = React.useRef<Cart>(loadedCart);

    const updateCartModel = (): void => {
        apiProxy.updateCartModel(new CartUpdateRequest(cartRef.current, countryName))
            .then((result) => {
                cartApi.saveNewCart(result);
                const loadedCart = cartApi.loadCart();
                cartRef.current = loadedCart;
                gridApi?.setRowData(cartRef.current.cartItems);
            })
            .catch(error => setError(error.response.data))
    }

    const getPostage = (): void => {
        apiProxy.getPostage(getSubTotal(), countryName)
            .then(result => setPostage(result))
            .catch(error => setError(error.response.data))
    }

    useEffect(() => {
        updateCartModel();
    }, [countryName]);

    useEffect(() => {
        getPostage();
    }, [cartRef.current, countryName]);

    const currencyFormatter = (params: any) => {
        if (params && params.data && params.data.product && params.data.product.price) return `${params.data.product.price.currencyCode} ${formatNumber(params.value)}`;
        return params.value;
    };

    const getAmount = (params: any): number => {
        if (params && params.data && params.data.product && params.data.product.price) return params.data.product.price.amount * params.data.count;
        return 0;
    }

    const onRemove = (): void => {
        setError(undefined);
        if (selectedProducts && selectedProducts.length > 0) {
            selectedProducts.map(x => {
                const removedCartItem: CartItem | undefined = cartRef.current.cartItems.find(z => z.product.productId == x);
                if (removedCartItem != undefined) cartRef.current.removeItemFromCartAll(removedCartItem.product);
            })
        }
        setSelectedProducts([]);
        const loadedCart = cartApi.loadCart();
        cartRef.current = loadedCart
        gridApi?.setRowData(cartRef.current.cartItems);
    }

    const onOrder = (): void => {
        setError(undefined);
        const date = new Date();
        const orderId = createGuid();
        apiProxy.placeOrder(new Order(orderId, date.toISOString(), cartRef.current, countryName))
            .then((result) => {
                if (result == true) {
                    sessionStorage.removeItem('shoppingCart');
                    sessionStorage.setItem('lastOrderId', orderId);
                    onOrderProcessed();
                }
                else {
                    setError("Failed to process the order, please try again");
                }
            })
            .catch(error => setError(error.response.data))
    }

    const gridOptions = {
        columnDefs: [
            { checkboxSelection: true, width: 50 },
            { field: 'product.productName', headerName: "Name", width: 400 },
            { field: 'product.price.amount', headerName: "Price", width: 200, cellStyle: { textAlign: 'right' }, valueFormatter: currencyFormatter },
            { field: 'count', headerName: "Count", width: 200, cellStyle: { textAlign: 'right' } },
            { headerName: "Amount", width: 200, cellStyle: { textAlign: 'right' }, valueGetter: getAmount, valueFormatter: currencyFormatter },
        ],
        defaultColDef: {
            sortable: true,
            wrapText: true,
            autoHeight: true,
        },
        rowSelection: 'multiple',
        onSelectionChanged: onSelectionChanged,
        onGridReady: onCartGridReady
    };

    function onCartGridReady(params: any): void {
        setGridApi(params.api);
    }

    function onSelectionChanged(evt: AgGridEvent) {
        setError(undefined);
        const selectedRows: CartItem[] = evt.api.getSelectedRows();
        const selectedProductIds: number[] = selectedRows.map(x => x.product.productId)
        setSelectedProducts(selectedProductIds);
    }

    const getSubTotal = (): number => {
        const subTotal = cartRef.current.cartItems.map(x => x.count * x.product.price.amount).reduce((a, b) => a + b, 0);
        return subTotal;
    }

    const getSubTotalStr = (): string => {
        const ccyCode = cartRef.current.cartItems.length > 0 ? cartRef.current.cartItems[0].product.price.currencyCode : "";
        return `${ccyCode} ${formatNumber(getSubTotal())}`;
    }

    const getPostageStr = (): string => {
        return postage == undefined ? "" : `${postage.currencyCode} ${formatNumber(postage.amount)}`;
    }

    const getTotalStr = (): string => {
        const ccyCode = cartRef.current.cartItems.length > 0 ? cartRef.current.cartItems[0].product.price.currencyCode : "";
        const shippingCost = postage == undefined ? 0 : postage.amount;
        const total = getSubTotal() + shippingCost;
        return `${ccyCode} ${formatNumber(total)}`;
    }

    return (
        <div>
            {error ? <div className="paddedUpDown"><Alert variant='outlined' severity='error'>Error: {error}</Alert></div> : null}
            <Button variant="contained" onClick={onRemove} disabled={!(selectedProducts && selectedProducts.length > 0)} >Remove from cart</Button>
            <Button variant="contained" onClick={onOrder} disabled={!(cartRef.current.cartItems.length > 0) || error != undefined} color="success" style={{ float: 'right' }}>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Place Order&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</Button>
            <div className="ag-theme-alpine paddedUpDown">
                <AgGridReact gridOptions={gridOptions} rowData={cartRef.current.cartItems} domLayout={'autoHeight'}></AgGridReact>
            </div>
            <div className='cartTotals' >Subtotal:&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;{getSubTotalStr()}</div>
            <div className='cartPostage' >Postage:&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;{getPostageStr()}</div>
            <div className='cartTotals' >Total:&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;{getTotalStr()}</div>
        </div>
    );
}