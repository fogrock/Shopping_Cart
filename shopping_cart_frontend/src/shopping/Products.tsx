import React from 'react';
import { ApiProxy, Product } from './ApiProxy';
import { useEffect } from 'react';
import Alert from '@mui/material/Alert';
import { AgGridReact } from 'ag-grid-react';
import { AgGridEvent, GridApi } from 'ag-grid-community';
import 'ag-grid-community/dist/styles/ag-grid.css';
import 'ag-grid-community/dist/styles/ag-theme-alpine.css';
import Button from '@mui/material/Button';
import { Cart, CartApi } from './CartApi';
import { formatNumber } from '../shared/Utils';
import '../App.css';

export interface ProductsProps {
    countryName: string
}

export function Products(props: ProductsProps) {
    const { countryName } = props;
    const [error, setError] = React.useState<string | undefined>(undefined);
    const [products, setProducts] = React.useState<Product[]>([]);
    const [selectedProducts, setSelectedProducts] = React.useState<number[]>([]);
    const [gridApi, setGridApi] = React.useState<GridApi | undefined>(undefined);
    const currencyFormatter = (params: any) => {
        if (params && params.data) return `${params.data.price.currencyCode} ${formatNumber(params.value)}`;
        return params.value;
    };

    const gridOptions = {
        columnDefs: [
            { checkboxSelection: true, width: 50 },
            { field: 'productName', headerName: "Name", width: 400 },
            { field: 'productDescription', headerName: "Description", width: 600, resizable: true, flex: 1 },
            { field: 'price.amount', headerName: "Price", width: 180, cellStyle: { textAlign: 'right' }, valueFormatter: currencyFormatter }
        ],
        defaultColDef: {
            sortable: true,
            wrapText: true,
            autoHeight: true,
        },
        rowSelection: 'multiple',
        onSelectionChanged: onSelectionChanged,
        onGridReady: onProductsGridReady,
    };

    function onProductsGridReady(params: any): void {
        setGridApi(params.api);
    }

    function onSelectionChanged(evt: AgGridEvent) {
        setError(undefined);
        const selectedRows: Product[] = evt.api.getSelectedRows();
        const selectedProductIds: number[] = selectedRows.map(x => x.productId)
        setSelectedProducts(selectedProductIds);
    }

    const getProductList = () => {
        const apiProxy = new ApiProxy();
        apiProxy.getProductList(countryName)
            .then(result => setProducts(result))
            .catch(error => setError(error.response.data))
    }

    const onAdd = () => {
        setError(undefined);
        const cartApi = new CartApi();
        const cart: Cart = cartApi.loadCart();
        if (selectedProducts && selectedProducts.length > 0) {
            selectedProducts.map(x => {
                const addedProduct: Product | undefined = products.find(z => z.productId == x);
                if (addedProduct != undefined) cart.addItemToCart(addedProduct, 1);
            })
        }
        setSelectedProducts([]);
        gridApi?.deselectAll();
    }

    useEffect(() => {
        getProductList();
    }, [countryName]);

    return (
        <div>
            {error ? <div className="paddedUpDown"><Alert variant='outlined' severity='error'>Error: {error}</Alert></div> : null}
            <div className="ag-theme-alpine paddedUpDown">
                <AgGridReact gridOptions={gridOptions} rowData={products} domLayout={'autoHeight'}></AgGridReact>
            </div>
            <Button variant="contained" onClick={onAdd} disabled={!(selectedProducts && selectedProducts.length > 0)} >Add to cart</Button>
        </div>
    );
}