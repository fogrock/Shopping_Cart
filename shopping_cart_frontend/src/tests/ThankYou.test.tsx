import React from 'react';
import { render, screen } from '@testing-library/react';
import { ThankYou } from '../shopping/ThankYou';

test('renders buttons and selector', () => {
    render(<ThankYou />);
    const linkElement1 = screen.getByText(/Thank you./i);
    expect(linkElement1).toBeInTheDocument();
    const linkElement2 = screen.getByText(/Your order was completed successfully/i);
    expect(linkElement2).toBeInTheDocument();
    const linkElement3 = screen.getByText(/order id:/i);
    expect(linkElement3).toBeInTheDocument();
});

it("matches snapshot", () => {
    const { asFragment } = render(<ThankYou />);
    expect(asFragment()).toMatchSnapshot();
});