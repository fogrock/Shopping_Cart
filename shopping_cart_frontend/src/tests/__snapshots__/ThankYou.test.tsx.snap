// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiTypography-root MuiTypography-h3 MuiTypography-gutterBottom css-1mktp8m-MuiTypography-root"
    >
      Thank you.
    </div>
    <div
      class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
    >
      Your order was completed successfully
    </div>
    <span
      class="MuiTypography-root MuiTypography-button MuiTypography-gutterBottom css-w5np7v-MuiTypography-root"
    >
      order id: 
    </span>
  </div>
</DocumentFragment>
`;
